presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1808,53 +1809,151 @@ private Scope computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope so
+                int fieldIndex = relationType.indexOf(field);
+                FieldReference fieldReference = new FieldReference(fieldIndex);
+                outputExpressionBuilder.add(fieldReference);
+                itemOutputFieldBuilder.add(Field.newUnqualified(field.getName(), field.getType(), field.getOriginTable(), field.getOriginColumnName(), false));
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(fieldReference, scope);
+
+                Type type = expressionAnalysis.getType(fieldReference);
+                if (node.getSelect().isDistinct() && !type.isComparable()) {
+                    throw new SemanticException(TYPE_MISMATCH, node.getSelect(), "DISTINCT can only be applied to comparable types (actual: %s)", type);
+                }
+            }
+            analysis.setSelectAllResultFields(allColumns, itemOutputFieldBuilder.build());
+        }
+
+        private void analyzeAllFieldsFromRowTypeExpression(Expression expression, AllColumns allColumns, QuerySpecification node, Scope scope, ImmutableList.Builder<Expression> outputExpressionBuilder)
+        {
+            ImmutableList.Builder<Field> itemOutputFieldBuilder = ImmutableList.builder();
+
+            ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, scope);
+            Type type = expressionAnalysis.getType(expression);
+            if (!(type instanceof RowType)) {
+                throw new SemanticException(INVALID_ALL_COLUMNS_TARGET, node.getSelect(), "SELECT * can only be applied to expression of Row type");
+            }
+            if (!allColumns.getAliases().isEmpty()) {
+                validateColumnAliases(allColumns.getAliases(), ((RowType) type).getFields().size());
+            }
+            analysis.recordSubqueries(node, expressionAnalysis);
+
+            for (int i = 0; i < ((RowType) type).getFields().size(); i++) {
+                Expression outputExpression = new SubscriptExpression(expression, new LongLiteral("" + (i + 1)));
+                ExpressionAnalysis outputExpressionAnalysis = analyzeExpression(outputExpression, scope);
+                outputExpressionBuilder.add(outputExpression);
+
+                Type outputExpressionType = outputExpressionAnalysis.getType(outputExpression);
+                if (node.getSelect().isDistinct() && !(outputExpressionType.isComparable())) {
+                    throw new SemanticException(TYPE_MISMATCH, node.getSelect(), "DISTINCT can only be applied to comparable types (actual: %s)", type.getTypeParameters().get(i));
+                }
+
+                Optional<String> oldName = ((RowType) type).getFields().get(i).getName();
+                if (!allColumns.getAliases().isEmpty()) {
+                    String name = (allColumns.getAliases().get(i)).getValue();
+                    itemOutputFieldBuilder.add(Field.newUnqualified(Optional.of(name), outputExpressionType, Optional.empty(), oldName, true));
```

Comment:
I don't think using `oldName` here is correct. The `originName` in `Field` is used for column access control checks, so it has to be a proper column in the source table.